flatten-data-ghcn_d.ipynb

```diff
@@ -0,0 +1,1167 @@
+{
```

Comment:
<p>A few questions/comments</p><p>1) This part looks like its a combinations of the previous cells above? Is there a reason why theres a lot of duplication</p><p>2)You may consider using a list comprehension to get the dates/tables you want and save it as a list to iterate over rather than running it in a try-except (much less error prone and you can separate logic detecting tables vs logic querying BQ)</p><p>3)Consider using f-strings for readability</p><p>4) Also it is best practice not have a bare exception as in </p><pre class="ql-syntax" spellcheck="false">try:
  do something
except:
  do something if failed
</pre><p> </p><p>A better practice would be</p><p> </p><pre class="ql-syntax" spellcheck="false">try:
  do something
except RuntimeError:
  do something if failed
</pre><p> </p><p>This is because it'll allow you take advantages of the control flow depending on the errors caught, for example the Bigquery client has several <a href="https://github.com/googleapis/google-cloud-python/blob/master/bigquery/google/cloud/bigquery/dbapi/exceptions.py" target="_blank">Exceptions</a>, so you can do something like</p><p> </p><pre class="ql-syntax" spellcheck="false">try:
  do something
except DataError:
  do something
except OperationalError:
  do something else
finally: -- optional
   do something regardless
</pre><p> </p><p>This is because when querying a DB, there could be errors you run into thats related to the query (in which you would want to stop and debug) vs an error with the service (in which you can set up a retry with a constant/exponential backoff).</p><p> </p><p>No need to dive too deep into it but just thought that you should know in case we move down the road.</p><p> </p><p>5) BQ returns a JSON/dict as a response from their API, do we need to instantiate it to a dataframe every time? Why not approach it like</p><p> </p><pre class="ql-syntax" spellcheck="false">results = []
for table_name in table_used:
  response = bq.query(query) # should return a json
  results.append(response)

df = pd.Dataframe.from_records(results)
</pre><p> </p><p>This way we only hold one data frame in memory once and the responses of the BQ queries can just be a list. That way we can save how much data is held in memory and also give it the opportunity for multithreading.</p><p> </p><p>6) Style nitpick. If you end your filename with <code>*.csv.gz</code>  it will automatically apply GZIP compression without explicitly having to pass in <code>compression='gzip'</code>  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/DataKind-DC/audubon-cbc/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>